### ISSUE TITLE
[QUESTION] TFT feature importance over time

### ISSUE THREAD

**Issue by MichaelVerdegaal** [unit8co/darts#2685]
I've been enjoying working with the TFT model in Darts, and i'm now in the process of exploring the explainability functionality. I would like to have the feature importances per timestep, but as far as i see the only type supported in Darts right now is the global importances, is this correct? 

I'm looking to create plots similar to the ones in [Neuralforecast's TFT](https://nixtlaverse.nixtla.io/neuralforecast/models.tft.html#2-2-variable-importances-over-time).

![Image](https://github.com/user-attachments/assets/f94f56a6-4897-420a-a10d-141d7ee307cd)

**Comment by dennisbader** [comment]
Hi @MichaelVerdegaal. Thanks for raising this issue and sorry for the late response. Yes, currently we do not support the feature importances over time.

I added it to our backlog :)

**Comment by dennisbader** [comment]
Sounds great that you got it working :) Would you like to contribute to this or share how you did it? We'd be happy to give this a go/

**Comment by MichaelVerdegaal** [comment]
> Sounds great that you got it working :) Would you like to contribute to this or share how you did it? We'd be happy to give this a go.

Sharing is easy yeah. The core part of it was this subclass, inheriting from Explain
```python
import numpy as np
import pandas as pd
from darts import TimeSeries
from darts.explainability import TFTExplainer, TFTExplainabilityResult
from darts.logging import raise_log
from darts.utils.utils import generate_index
from matplotlib import pyplot as plt


class ExtendedTFTExplainer(TFTExplainer):
    """Extends default TFTExplainer class.

    The primary reason for doing so is the capability to get the feature importance
    weights per timestep, as the default explainer aggregates them for you.This also
    allows us to create custom explainability plots, more suitable for our
    use-case.
    """

    def explain(self, *args, **kwargs):
        """Override explain method to store sparse weights"""
        result = super().explain(*args, **kwargs)

        # Store raw weights from the last forward pass
        self._raw_decoder_weights = (
            self.model.model._decoder_sparse_weights.detach().cpu().numpy()
        )
        self._raw_encoder_weights = (
            self.model.model._encoder_sparse_weights.detach().cpu().numpy()
        )
        return result

    def feat_importance_per_timestep(self, feat_type: str = "encoder") -> pd.DataFrame:
        """Get feature importance per timestep

        Args:
            feat_type: type of importance, encoder or decoder

        Returns:
            DataFrame with importances
        """
        if not hasattr(self, "_raw_decoder_weights") and not hasattr():
            raise ValueError("Sparse weights not found, run explain() first")

        match feat_type:
            case "encoder":
                weights = self._raw_encoder_weights
                variable_names = self.model.model.encoder_variables
            case "decoder":
                weights = self._raw_decoder_weights
                variable_names = self.model.model.decoder_variables
            case _:
                raise ValueError("feat_type must be 'encoder' or 'decoder'")

        # weights shape: [batch, timesteps, attention heads, features]
        weights = np.mean(weights, axis=(0, 2)) * 100  # shape to: [timesteps, features]

        # convert raw feature names to mapped names
        feature_names = [self._name_mapping[name] for name in variable_names]

        return pd.DataFrame(
            weights,
            columns=feature_names,
            index=[f"step_{i + 1}" for i in range(weights.shape[0])],
        )
``` 

After which i created a new plot function. My basis of this was a mix of plot_variable_selection() and plot_attention().
```python
    def plot_importance_over_time(
            self,
            expl_result: TFTExplainabilityResult,
            feat_type: str = "encoder",
            normalize: bool = False,
            show_index_as: str = "relative",
            ax: plt.Axes | None = None,
            show_plot: bool = True,
            title_suffix: str = "",
    ) -> tuple[plt.Figure, plt.Axes]:
        """Plots feature importance over time with merged cyclical features

        Args:
            expl_result: explainability result
            feat_type: type of importance, encoder (past) or decoder (future)
            normalize: if true normalize by mean attention
            show_index_as: show index as time, or number relative to prediction point
            ax: axis to plot on
            show_plot: whether to show the plot or not
            title_suffix: suffix to add to title, mostly used for filepaths

        Returns:
            matplotlib figure and axis
        """
        # Retrieve components
        attention = expl_result.get_explanation(component="attention")
        importances = self.feat_importance_per_timestep(feat_type)

        # Rename columns to be more readable
        colnames = importances.columns
        colnames = [
            c.removesuffix("_futcov")
            .removesuffix("_pastcov")
            .removeprefix("darts_enc_fc_cyc_")
            for c in colnames
        ]
        importances.columns = colnames

        # Merge _sin and _cos columns
        base_features = set()
        for col in importances.columns:
            if col.endswith("_sin") or col.endswith("_cos"):
                base_name = col[:-4]  # Remove _sin or _cos
                base_features.add(base_name)

        merged_importances = pd.DataFrame(index=importances.index)
        for feat in base_features:
            sin_col = f"{feat}_sin"
            cos_col = f"{feat}_cos"

            # Only merge if both sin and cos columns exist
            if sin_col in importances.columns and cos_col in importances.columns:
                merged_importances[feat] = importances[sin_col] + importances[
                    cos_col]
            elif sin_col in importances.columns:
                merged_importances[feat] = importances[sin_col]
            elif cos_col in importances.columns:
                merged_importances[feat] = importances[cos_col]

        # Add non-cyclic columns
        for col in importances.columns:
            if not (col.endswith("_sin") or col.endswith("_cos")):
                merged_importances[col] = importances[col]

        # Use merged_importances instead of importances for plotting
        importances = merged_importances

        if ax is None:
            fig, ax = plt.subplots(figsize=(20, 10), layout="constrained")

        # Prepare x-axis
        match show_index_as:
            case "relative":
                x_ticks = generate_index(
                    start=-self.model.input_chunk_length, end=self.n - 1
                )
                attention = TimeSeries.from_times_and_values(
                    times=generate_index(
                        start=-self.model.input_chunk_length, end=self.n - 1
                    ),
                    values=attention.values(copy=False),
                    columns=attention.components,
                )
                x_label = "Index relative to first prediction point"
            case "time":
                x_ticks = attention.time_index
                x_label = "Time index"
            case _:
                raise_log(
                    ValueError(
                        "`show_index_as` must either be 'relative', or 'time'.")
                )

        # Slice mean attention based on feat_type
        mean_attention: TimeSeries = attention.mean(1)
        split_point = self.n
        plot_title = f"{title_suffix} "
        if feat_type == "encoder":
            mean_attention = mean_attention[:-split_point]
            x_ticks = x_ticks[:-split_point]
            plot_title += "past variable importance over time"
        elif feat_type == "decoder":
            mean_attention = mean_attention[-split_point:]
            x_ticks = x_ticks[-split_point:]
            plot_title += "future variable importance over time"

        # Normalize and plot attention if applicable
        if normalize:
            importances = importances.divide(100)  # remove style percentages
            importances = importances.multiply(
                mean_attention.univariate_values(), axis=0
            )
            ax.plot(
                x_ticks,
                mean_attention.univariate_values(),
                color="black",
                linestyle="-",
                marker="o",
                linewidth=1,
                alpha=0.5,
                label="mean_attention",
            )

        # Plot stacked bars with feature importances
        colors = plt.cm.tab20(np.linspace(0, 1, 20))
        bottom = np.zeros(len(importances))
        for idx, col in enumerate(importances.columns):
            # Use hatching only after the first 20 features
            pattern = "//" if idx >= 20 else None
            color_idx = idx % 20  # Cycle through the colors

            _ = ax.bar(
                x_ticks,
                importances[col].values,
                0.6,
                label=col,
                bottom=bottom,
                color=colors[color_idx],
                hatch=pattern,
                edgecolor="black",
                linewidth=0.5,
            )
            bottom += importances[col]  # Stack bars

        # Adjust plot
        ax.set_xlabel(x_label)
        ax.set_ylabel("Attention")
        if normalize:
            plot_title += ", normalized by attention"
        plt.title(plot_title)

        # Adjust legend
        ax.legend(bbox_to_anchor=(1.05, 1), loc="upper left", borderaxespad=0.0,
                  ncol=1)

        # Rotate x-axis labels if they're dates
        if show_index_as == "time":
            plt.xticks(rotation=45)
        if show_plot:
            plt.show()

        return fig, ax
```

I don't think the plotting method is a good example to follow as it was written during overtime, but i've been using this in production with my model and have some thoughts:
- Readable visualization is definitely a concern. Above 20 features it becomes near impossible to read. This gets worse the bigger your input chunk length / output chunk length is (120 / 42 for me).
- A large part of this was the cyclical time features i added via the Darts add_encoders function. I decided to merge these together for my code, which made it much more readable, and saved me from having to explain what the _sin/_cos suffix is constantly. This means that perhaps it should be possible to allow users to filter to specific features more directly, so those can be inspected individually.
-The "normalized by attention" functionality (see Nixtla again) i honestly didn't see the use of. I'd rather just have a good attention plot next to the feature importances plot. The reasoning for this is that the bars become near impossible to read when attention is low, and also when attention is high it's not possible to see the per timestep importance change, which i'd argue is the entire strength of what we want to do here.

All in all i quite liked what i tried so far. For example, i had some cases where the importances had a massive value shift at some timestep, which prompted a closer look at my source data, leading to the discovery of a distribution shift i didn't notice.

**Comment by exactml** [comment]
@dennisbader still interested in this?

**Comment by dennisbader** [comment]
@exactml, yes we're still interested in this :) Would you like to contribute? The goal would be to have the logic covered in the existing `TFTExplainer` (e.g. not via a subclass). Regarding the output, it would also be good the leverage our `TimeSeries` instead of a `pandas.DataFrame`. What do you think?

**Comment by exactml** [comment]
@dennisbader Thanks for the quick response, happy to contribute this.

I've split the work into two phases so the data-layer change can be reviewed on its own before we lock in a plotting API:

1. Data layer (this PR): adds per-timestep (non-aggregated) encoder/decoder variable importances to the existing TFTExplainer, following both constraints from your last comment — no subclass, and output as TimeSeries via TFTExplainabilityResult.get_encoder_importance_over_time() / get_decoder_importance_over_time().

2. Plotting (follow-up): a dedicated visualization for the importance over time. Holding off on this until the data API above is settled — partly because @MichaelVerdegaal's own retrospective earlier in this thread flagged real open questions there (readability past ~20 features, whether to auto-merge cyclic sin/cos pairs, and that the "normalize by attention" idea from Nixtla didn't actually pull its weight for him). Feels like it deserves its own discussion rather than getting bundled into this PR.

Let me know if you'd rather see both in one PR instead, or if the phased approach works for you. Also very open to feedback on the API shape before this goes further.

**Comment by exactml** [comment]
Hi @dennisbader, following up on the plotting side of this — opened [PR #3174](https://github.com/unit8co/darts/pull/3174).

TFTExplainer.plot_variable_selection_over_time() plots the per-timestep encoder/decoder variable importances from #3170 as line charts (one line per variable) rather than the stacked-bar approach from the earlier prototype in this thread, specifically to sidestep the axis-crowding failure mode stacked bars hit on long input_chunk_length.

**Default (show_index_as="relative"):**

<img width="928" height="701" alt="Image" src="https://github.com/user-attachments/assets/d576d279-0963-4f1a-91e1-7e513037ce18" />


**show_index_as="time" — same data, actual calendar dates instead:**

<img width="938" height="699" alt="Image" src="https://github.com/user-attachments/assets/e549bb46-24d4-4d18-b777-46fbfc16d8d8" />

**Stress test — the small examples above don't really exercise the concerns raised earlier in this thread, so I also tested with a 60-day input / 20-day forecast window and 36/18 available encoder/decoder variables (vs. the default max_nr_components=10 cap). Even at that scale, the plot stays readable — variables capped to 10 lines per subplot, and matplotlib's date locator keeps the x-axis to a handful of clean ticks instead of one label per day:**

<img width="915" height="753" alt="Image" src="https://github.com/user-attachments/assets/9a89e9fc-6c5e-4e70-bb67-3e39e584dca9" />
